### PR TITLE
Record the seed and random states during training

### PR DESCRIPTION
**Pull request by davidwyld**
This PR will:
 - During training record:
   - Record in a pickle file:
     - the seed value set by the user
     - the random state of numpy
     - the random state of python's random
 - Fix a bug in the seed setting routine in `__main__.py` which caused an error if the seed was omitted from the config file or command line.

### CODE REVIEW

**Review by drj11 — APPROVED** [review]
fine.

**Comment by davidwyld** [comment]
Have updated this to write to JSON instead so the file is human readable (77a1c70)

File: train.py [hunk 1]
```diff
@@ -294,6 +294,29 @@ def buildCallbacks(params, model, dataset):
...
+    import random
+    import json
+
+    n = numpy.random.get_state()
+    p = random.getstate()
+
+    n = list(numpy.random.get_state())
+    n[1] = n[1].tolist()
+
+    p = list(numpy.random.get_state())
```
Comment (drj11) [inline train.py]: lol no.

You accidentally overwrite `p` which is supposed to be the Python random state.

**Review by drj11 — CHANGES_REQUESTED** [review]
fix bug!

File: train.py (hunk 1 above)
Comment (davidwyld) [reply train.py]: Thanks for the spot!

**Review by fredblain — APPROVED** [review]

File: train.py [hunk 2]
```diff
@@ -299,14 +299,11 @@ def save_random_states(write_path, user_seed=None):
...
-    n = numpy.random.get_state()
-    p = random.getstate()
-
     n = list(numpy.random.get_state())
     n[1] = n[1].tolist()
 
-    p = list(numpy.random.get_state())
-    p[1] = p[1].tolist()
+    p = list(random.getstate())
+    p[1] = list(p[1])
```
Comment (drj11) [inline train.py]: this line is not needed, but mostly harmless (for the builtin `random` module, `.getstate()` already returns something that can be JSON serialised.

**Review by drj11 — APPROVED** [review]

**Pull request merged**